<!DOCTYPE html>
<html lang="en" data-bs-theme="light">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="../img/favicon.ico">
        <title>Tutorial - seabreeze</title>
        <link href="../css/bootstrap.min.css" rel="stylesheet">
        <link href="../css/fontawesome.min.css" rel="stylesheet">
        <link href="../css/brands.min.css" rel="stylesheet">
        <link href="../css/solid.min.css" rel="stylesheet">
        <link href="../css/v4-font-face.min.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link id="hljs-light" rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" >
        <link id="hljs-dark" rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github-dark.min.css" disabled>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script>hljs.highlightAll();</script> 
    </head>

    <body>
        <div class="navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
            <div class="container">
                <a class="navbar-brand" href="..">seabreeze</a>
                <!-- Expander button -->
                <button type="button" class="navbar-toggler" data-bs-toggle="collapse" data-bs-target="#navbar-collapse" aria-controls="navbar-collapse" aria-expanded="false" aria-label="Toggle navigation">
                    <span class="navbar-toggler-icon"></span>
                </button>

                <!-- Expanded navigation -->
                <div id="navbar-collapse" class="navbar-collapse collapse">
                        <!-- Main navigation -->
                        <ul class="nav navbar-nav">
                            <li class="nav-item">
                                <a href=".." class="nav-link">Home</a>
                            </li>
                            <li class="nav-item">
                                <a href="../installation/" class="nav-link">Installation</a>
                            </li>
                            <li class="nav-item">
                                <a href="../usage/" class="nav-link">Usage</a>
                            </li>
                            <li class="nav-item">
                                <a href="../output/" class="nav-link">Output</a>
                            </li>
                            <li class="nav-item">
                                <a href="./" class="nav-link active" aria-current="page">Tutorial</a>
                            </li>
                        </ul>

                    <ul class="nav navbar-nav ms-md-auto">
                        <li class="nav-item">
                            <a href="#" class="nav-link" data-bs-toggle="modal" data-bs-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li class="nav-item">
                                <a rel="prev" href="../output/" class="nav-link">
                                    <i class="fa fa-arrow-left"></i> Previous
                                </a>
                            </li>
                            <li class="nav-item">
                                <a rel="next" class="nav-link disabled">
                                    Next <i class="fa fa-arrow-right"></i>
                                </a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
            <div class="row">
                    <div class="col-md-3"><div class="navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
    <div class="navbar-header">
        <button type="button" class="navbar-toggler collapsed" data-bs-toggle="collapse" data-bs-target="#toc-collapse" title="Table of Contents">
            <span class="fa fa-angle-down"></span>
        </button>
    </div>

    
    <div id="toc-collapse" class="navbar-collapse collapse card bg-body-tertiary">
        <ul class="nav flex-column">
            
            <li class="nav-item" data-bs-level="1"><a href="#tutorial" class="nav-link">Tutorial</a>
              <ul class="nav flex-column">
            <li class="nav-item" data-bs-level="2"><a href="#analyse-genome-sizes" class="nav-link">Analyse genome sizes</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#find-insertion-sequences" class="nav-link">Find insertion sequences</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#predict-svs" class="nav-link">Predict SVs</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#annotate-genes-in-svs" class="nav-link">Annotate genes in SVs</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#predict-sv-mechanism" class="nav-link">Predict SV mechanism</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#predict-replichore-and-inversion-balance" class="nav-link">Predict replichore and inversion balance</a>
              <ul class="nav flex-column">
              </ul>
            </li>
              </ul>
            </li>
        </ul>
    </div>
</div></div>
                    <div class="col-md-9" role="main">

<h1 id="tutorial">Tutorial</h1>
<p>This is a tutorial with example data for <em>seabreeze</em>. Please <a href="../installation/">install the software, create the conda enviroment and activate it</a> first. </p>
<p>To get started, navigate to the root of the <em>seabreeze</em> directory, and copy the <code>example/</code> directory:</p>
<pre><code>cp -r example/ data/
</code></pre>
<p>The newly created <code>data/</code> folder should look like this:</p>
<pre><code>|---data/
|   |
|   |---data.csv
|   |---ori_dif_sequences.csv
|   |---02_genomes/
|   |   |
|   |   |---REL606_evolved_1.fasta
|   |   |---REL606_evolved_2.fasta
|   |   |---REL606.fasta
</code></pre>
<p>For this tutorial, structural variant mutations have been simulated on the <em>E. coli</em> strain REL606 genome using <em>breseq</em>. These mutated genome assemblies are <code>REL606_evolved_1.fasta</code> and <code>REL606_evolved_2.fasta</code>. We aim to identify the structural variants in these assemblies relative to their ancestor, <code>REL606.fasta</code>. These pairwise comparisons to be performed are specified in a csv file called <code>data.csv</code> in the <code>data/</code> directory.  For more information about the valid format of this file, please see the <a href="../usage/">usage</a> page.</p>
<p>To view this csv file:</p>
<pre><code>cat data/data.csv
</code></pre>
<p>It should look like this:</p>
<table>
<thead>
<tr>
<th>assembly</th>
<th>ancestor</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td>REL606</td>
<td>REL606</td>
<td></td>
</tr>
<tr>
<td>REL606_evolved_1</td>
<td>REL606</td>
<td></td>
</tr>
<tr>
<td>REL606_evolved_2</td>
<td>REL606</td>
<td></td>
</tr>
</tbody>
</table>
<p>We include a comparison of <code>REL606</code> to itself as a quality check step, as we do not expect any mutations when a genome is compared to itself.</p>
<p>All of the following commands should be run from the <em>seabreeze</em> root directory. Here, 4 cores have been allocated to run <em>seabreeze</em> but this number can be higher on lower depending on your hardware. </p>
<h2 id="analyse-genome-sizes">Analyse genome sizes</h2>
<p>This command compares the sizes of the assemblies to their specified ancestors in a pairwise manner. </p>
<pre><code>snakemake --use-conda --cores 4 analyse_genome_sizes
</code></pre>
<p>The output file generated is <code>data/04_rename_genome/genome_size_stats.csv</code>. Please see <a href="../output/">output</a> for more information about the fields in this table. This table should look this this:</p>
<table>
<thead>
<tr>
<th>assembly</th>
<th>ancestor</th>
<th>size_assembly</th>
<th>size_ancestor</th>
<th>difference</th>
<th>percent_change</th>
</tr>
</thead>
<tbody>
<tr>
<td>REL606</td>
<td>REL606</td>
<td>4629812</td>
<td>4629812</td>
<td>0</td>
<td>0.0</td>
</tr>
<tr>
<td>REL606_evolved_1</td>
<td>REL606</td>
<td>4617111</td>
<td>4629812</td>
<td>-12701</td>
<td>-0.2743</td>
</tr>
<tr>
<td>REL606_evolved_2</td>
<td>REL606</td>
<td>4549910</td>
<td>4629812</td>
<td>-79902</td>
<td>-1.7258</td>
</tr>
</tbody>
</table>
<p>As we expected, there is no size difference when <code>REL606</code> is compared to itself. We can see that both of the assemblies <code>REL606_evolved_1</code> and <code>REL606_evolved_2</code> are smaller than the ancestor.</p>
<h2 id="find-insertion-sequences">Find insertion sequences</h2>
<p>This command uses <a href="https://github.com/xiezhq/ISEScan/blob/master/README.md?plain=1">ISEscan</a> to annotate partial and full-length IS elements in the genomes. </p>
<pre><code>snakemake --use-conda --cores 4 predict_IS_elements
</code></pre>
<p>All of the output generated is stored in <code>data/05_isescan_tables/</code>. This tutorial should generate the following files in this directory:</p>
<pre><code>|---data/
|   |---05_isescan_tables/
|   |   |
|   |   |---REL606.csv
|   |   |---REL606_evolved_1.csv
|   |   |---REL606_evolved_2.csv
|   |   |---REL606/
|   |   |---REL606_evolved_1/
|   |   |---REL606_evolved_2/
</code></pre>
<p>Each assembly has a .csv file that describes the locations of IS elements in that genome, and a directory where the additionally files generated by <em>ISEScan</em> are made. For information about interpreting this csv file, we refer users to the official <a href="https://github.com/xiezhq/ISEScan/blob/master/README.md">ISEScan documentation</a>.</p>
<h2 id="predict-svs">Predict SVs</h2>
<p>This command aligns the assembly-ancestor pairs, predicts structural variants and generates synteny plots. </p>
<pre><code>snakemake --use-conda --cores 4 predict_structural_variants
</code></pre>
<p>This will generate the following directories:</p>
<pre><code>|---data/
|   |---06_nucmer_alignment/
|   |   |
|   |   |---REL606/
|   |   |---REL606_evolved_1/
|   |   |---REL606_evolved_2/
|   |   |
|   |---07_syri_output/
|   |   |
|   |   |---REL606/
|   |   |---REL606_evolved_1/
|   |   |---REL606_evolved_2/
</code></pre>
<p>We are primarily interested in a table describing the SVs and the synteny plot. This command also generates other files that may be useful, but are not required. Please see <a href="../output/">output</a> for information about those files.</p>
<p>Let's look at the SVs for <code>REL606_evolved_1</code></p>
<pre><code>cd data/07_syri_output/REL606_evolved_1
cat REL606_evolved_1_clean.syri.out | cut --complement -f 4-5
</code></pre>
<p>This command prints the tsv file <code>REL606_evolved_1_clean.syri.out</code> to the terminal, while hiding the 4th and 5th columns. These columns show the reference/query sequences for indels, and hence can be quite large. Note that older versions of the UNIX <code>cut</code> command may not support the <code>--complement</code> option, and you may need to either update <code>cut</code> or use other UNIX commands to hide the 4th and 5th columns. You can also view this file with a spreadsheet application, as it a tsv file. </p>
<pre><code>genome  1       590471  genome  1       595344  SYN1    -   SYN -
genome  590473  599656  genome  595344  595344  DEL -   DEL -
genome  599657  700001  genome  595345  697137  SYN5    -   SYN -
genome  699997  2100900 genome  697133  2105267 INV15   -   INV -
genome  2100896 2775884 genome  2105263 2784492 SYN7    -   SYN -
genome  2775970 2789333 genome  2784492 2784492 DEL -   DEL -
genome  2789334 3697156 genome  2784487 3686254 SYN11   -   SYN -
genome  3002909 3008938 genome  2998062 2998062 DEL6    SYN11   DEL -
genome  3697154 4415713 genome  3687698 4400119 SYN12   -   SYN -
genome  3894996 3901133 genome  3885540 3885540 DEL7    SYN12   DEL -
genome  4415710 4629812 genome  4401562 4617111 SYN13   -   SYN -
genome  4615673 4629812 genome  4602972 4617111 SYN14   -   SYN -

</code></pre>
<p>We refer the user to the official <a href="https://schneebergerlab.github.io/syri/fileformat.html">SyRI documentation</a> for information on interpreting this table. You may notice that the reference and query chromosome ID may read as "genome". This is because SyRI attempts to call variants between homologous chromosomes, and hence expects the reference and query ID to be the same string. To prevent this error from occurring, the genomes are renamed to a placeholder string 'genome'. </p>
<p>Let us view the synteny plot  <code>REL606_evolved_1.plot.2.pdf</code> in this same directory. It can be opened with any application that allows you to view a PDF. </p>
<p><img alt="plot_1" src="../synteny_plot_1.png" /></p>
<p>The ancestor genome is the blue line, and the orange line is the evolved genome REL606_evolved_1. The grey ribbons show syntenic regions. We can see that there was one large inversion (in yellow) and several smaller deletions across the genome (in red).</p>
<p>One of the steps that <em>seabreeze</em> performs is to remove false positive SV calls and misassigned SVs. As an example of what this synteny plot looks like before this processing, we can view the plot <code>REL606_evolved_1.plot.pdf</code> </p>
<p><img alt="plot_2" src="../synteny_plot_2.png" /></p>
<p>Two deletions (~0.5Mb and ~2.8Mb) were not correctly called. There also appear to be many small "duplications". These correspond to spurious alignments between the many copies of IS elements, and do not always represent bona fide instances of structural variant mutations. It should also be noted that <em>seabreeze</em> does not track IS transposition events.</p>
<h2 id="annotate-genes-in-svs">Annotate genes in SVs</h2>
<p>To annotate the genes present in the deletions, inversions and amplifications, run this command. Make sure you navigate back to the <em>seabreeze</em> root directory before that.</p>
<pre><code>snakemake --use-conda --cores 4 annotate_SV_regions
</code></pre>
<p>The primary output generated by this command are a series of html files in </p>
<pre><code>|---data/
|   |---12_genome_diff_tables
|   |   |---html
|   |   |   |---REL606.html
|   |   |   |---REL606_evolved_1.html
|   |   |   |---REL606_evolved_2.html
</code></pre>
<p>Let us look at <code>REL606_evolved_1.html</code> as an example. This file can be opened by your web browser application.
 <img alt="html_1" src="../html_table_1.png" /></p>
<p>This table describes the mutations, their type, location and the genes present in them. For more information on this table, please see <a href="https://gensoft.pasteur.fr/docs/breseq/0.35.0/output.html#html-human-readable-output">breseq documentation</a>. </p>
<h2 id="predict-sv-mechanism">Predict SV mechanism</h2>
<p>Most structural variant mutations occur through recombination between homologous sites and insertion sequences in particular are often involved. This command annotates the insertion sequences at the boundaries of structural variants (for deletions and inversions only) and predicts putative mechanisms.</p>
<pre><code>snakemake --use-conda --cores 4 predict_SV_mechanism
</code></pre>
<p>This command generates several files: two summary files <code>deletion_mechanism.csv</code> and <code>inversion_mechanism.csv</code>, and three files for each of the three assemblies:</p>
<pre><code>|---data/
|   |---11_annotated_boundaries
|   |   |---deletion_mechanism.csv
|   |   |---inversion_mechanism.csv
|   |   |---REL606_boundaries.csv
|   |   |---REL606_deletion.csv
|   |   |---REL606_inversion.csv
|   |   |---REL606_evolved_1_boundaries.csv
|   |   |---REL606_evolved_1_deletion.csv
|   |   |---REL606_evolved_1_inversion.csv
|   |   |---REL606_evolved_2_boundaries.csv
|   |   |---REL606_evolved_2_deletion.csv
|   |   |---REL606_evolved_2_inversion.csv
</code></pre>
<p>Let's look at the output for the assembly <code>REL606_evolved_1_deletion.csv</code>, which describes the putative mechanism of the deletions:</p>
<table>
<thead>
<tr>
<th>ref_start</th>
<th>ref_stop</th>
<th>query_start</th>
<th>query_stop</th>
<th>tag_3</th>
<th>L_ref</th>
<th>L_ref_distance</th>
<th>R_ref</th>
<th>R_ref_distance</th>
<th>L_query</th>
<th>L_query_distance</th>
<th>R_query</th>
<th>R_query_distance</th>
<th>Mechanism</th>
<th>Evidence</th>
</tr>
</thead>
<tbody>
<tr>
<td>590473</td>
<td>599656</td>
<td>595344</td>
<td>595344</td>
<td>DEL</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td>other</td>
<td>NA</td>
</tr>
<tr>
<td>2775970</td>
<td>2789333</td>
<td>2784492</td>
<td>2784492</td>
<td>DEL</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td>IS3_61</td>
<td>-7</td>
<td>IS3_61</td>
<td>7</td>
<td>IS_mediated</td>
<td>evolved</td>
</tr>
<tr>
<td>3002909</td>
<td>3008938</td>
<td>2998062</td>
<td>2998062</td>
<td>DEL</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td>other</td>
<td>NA</td>
</tr>
<tr>
<td>3894996</td>
<td>3901133</td>
<td>3885540</td>
<td>3885540</td>
<td>DEL</td>
<td>IS3_61</td>
<td>0</td>
<td></td>
<td>0</td>
<td>IS3_61</td>
<td>0</td>
<td>IS3_61</td>
<td>0</td>
<td>IS_mediated</td>
<td>full</td>
</tr>
</tbody>
</table>
<p>We see that two deletions seem to involve IS3_61. For more information about this table, please see the <a href="../output/">output documentation</a>.</p>
<p>Now let us look at the output for the assembly <code>REL606_evolved_1_inversion.csv</code>, which describes the putative mechanism of the inversions:</p>
<table>
<thead>
<tr>
<th>ref_start</th>
<th>ref_stop</th>
<th>query_start</th>
<th>query_stop</th>
<th>tag_3</th>
<th>L_ref</th>
<th>L_ref_distance</th>
<th>R_ref</th>
<th>R_ref_distance</th>
<th>L_query</th>
<th>L_query_distance</th>
<th>R_query</th>
<th>R_query_distance</th>
<th>Mechanism</th>
<th>Evidence</th>
</tr>
</thead>
<tbody>
<tr>
<td>699997</td>
<td>2100900</td>
<td>697133</td>
<td>2105267</td>
<td>INV</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td></td>
<td>0</td>
<td>other</td>
<td>NA</td>
</tr>
</tbody>
</table>
<p>There was only a single inversion in this assembly, which occurred by an unknown mechanism <em>seabreeze</em> could not predict.</p>
<p>Looking at the summary file <code>deletion_mechnism.csv</code>, we can at a glance look at the mechanism for deletions in all of the assemblies:</p>
<table>
<thead>
<tr>
<th>clone</th>
<th>total</th>
<th>between_IS</th>
<th>IS_mediated</th>
<th>other</th>
</tr>
</thead>
<tbody>
<tr>
<td>REL606</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>REL606_evolved_2</td>
<td>4</td>
<td>0</td>
<td>1</td>
<td>3</td>
</tr>
<tr>
<td>REL606_evolved_1</td>
<td>4</td>
<td>0</td>
<td>2</td>
<td>2</td>
</tr>
</tbody>
</table>
<p>Similarly, for the inversions in <code>inversion_mechanism.csv</code>:</p>
<table>
<thead>
<tr>
<th>clone</th>
<th>total</th>
<th>between_IS</th>
<th>IS_mediated</th>
<th>other</th>
</tr>
</thead>
<tbody>
<tr>
<td>REL606</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>REL606_evolved_2</td>
<td>2</td>
<td>1</td>
<td>0</td>
<td>1</td>
</tr>
<tr>
<td>REL606_evolved_1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>1</td>
</tr>
</tbody>
</table>
<h2 id="predict-replichore-and-inversion-balance">Predict replichore and inversion balance</h2>
<p>For bacterial genomes that have a single origin and terminus, the genome can be divided into two replichores (halves) demarcated by the origin-terminus axis. It can be useful to know how the lengths of these two replichores differ between the ancestor-assembly pairs. In this figure, the two replichores have been depicted in yellow and green.</p>
<p><img alt="inversion_replichore_1" src="../inversion_replichore_1.png" /> </p>
<p>The origin-terminus axis can also be used to classify inversions as inter-replichore (if they occur across the axis) or intra-replichore if they are contained within a single replichore. This figure depicts an inter-replichore inversion across the origin, and an intra-replichore inversion within the green replichore.</p>
<p><img alt="inversion_replichore_2" src="../inversion_replichore_2.png" /></p>
<p>For inter-replichore inversions, we can describe the symmetry of the inversion across the axis. Asymmetric inversions can cause the length of the two replichores to change.</p>
<p><img alt="inversion_replichore_3" src="../inversion_replichore_3.png" /></p>
<p>This step requires user input (in addition to <code>data/data.csv</code>) to specify the sequences of the origin and terminus. For the origin, we recommend the <em>oriC</em> sequence, and for the terminus, we recommend the <em>dif</em> sequence. <em>seabreeze</em> looks for these sequences (or the reverse complement) in the genome for an exact and unique match, and will cause an error if an exact match is not found, or more than one exact match is found. There is no minimum/maximum length requirements for the supplied <em>oriC</em> and <em>dif</em> sequences as long as they meet the above criteria. <em>seabreeze</em> requires these two sequences for each ancestor, and assumes that the sequences have not mutated in the corresponding ancestor. This information is specified in a csv file in <code>data/ori_dif_sequences.csv</code>. </p>
<p>Let us view this file for our example:</p>
<pre><code>cat data/ori_dif_sequences.csv
</code></pre>
<p>We should be able to see this table:</p>
<table>
<thead>
<tr>
<th>ancestor</th>
<th>ori</th>
<th>dif</th>
</tr>
</thead>
<tbody>
<tr>
<td>REL606</td>
<td>GGATCCTGGGTATTAAAA</td>
<td>TCTTCCTTGGTTTATATT</td>
</tr>
</tbody>
</table>
<p>These sequences are the first few bases of the <em>oriC</em> and <em>dif</em> loci for <em>E. coli</em>. More information about what sequences are acceptable in this table are on the <a href="../usage/">usage</a> page. </p>
<p>We can now run the command from the <em>seabreeze</em> root directory:</p>
<pre><code>snakemake --use-conda --cores 4 predict_replichore_balance
</code></pre>
<p>This step generates several output files. First, let us look at the file that describes the lengths of the replichores in the assemblies <code>data/08_reindex_genome_oric/replichore_arms.csv</code>.</p>
<table>
<thead>
<tr>
<th>clone</th>
<th>ori</th>
<th>dif</th>
<th>length</th>
<th>arm_1</th>
<th>arm_2</th>
<th>ratio</th>
<th>percent</th>
</tr>
</thead>
<tbody>
<tr>
<td>REL606</td>
<td>0.0</td>
<td>2311092.0</td>
<td>4629812.0</td>
<td>2311092.0</td>
<td>2318720.0</td>
<td>1.003</td>
<td>50.082</td>
</tr>
<tr>
<td>REL606_evolved_1</td>
<td>0.0</td>
<td>1972601.0</td>
<td>4617111.0</td>
<td>1972601.0</td>
<td>2644510.0</td>
<td>1.341</td>
<td>57.276</td>
</tr>
<tr>
<td>REL606_evolved_2</td>
<td>0.0</td>
<td>1708781.0</td>
<td>4549910.0</td>
<td>1708781.0</td>
<td>2841129.0</td>
<td>1.663</td>
<td>62.444</td>
</tr>
</tbody>
</table>
<p>For more information, about this table, please see the <a href="../output/">output</a> page. <code>arm_1</code> and <code>arm_2</code> describe the length of the two replichores, and <code>percent</code> describes the percent of the total length of the genome that is in the longer replichore. This value is 50 for a perfectly balanced genome, and increases with increasing imbalance. We can see that both <code>REL606_evolved_1</code> and <code>REL606_evolved_2</code> are less balanced than <code>REL606</code>. </p>
<p>Now, let's look at the classification of the inversions in the assemblies in <code>data/11_annotated_boundaries/inversion_replichores_long.csv</code>. This file contains a list of all of the inversions across all of the assemblies.</p>
<table>
<thead>
<tr>
<th>clone</th>
<th>classification</th>
<th>length</th>
<th>mechanism</th>
<th>symmetry_percent</th>
</tr>
</thead>
<tbody>
<tr>
<td>REL606_evolved_2</td>
<td>across_dif</td>
<td>1500794</td>
<td>other</td>
<td>75.52768734416581</td>
</tr>
<tr>
<td>REL606_evolved_2</td>
<td>across_ori</td>
<td>1070041</td>
<td>between_IS</td>
<td>57.700686235387245</td>
</tr>
<tr>
<td>REL606_evolved_1</td>
<td>across_dif</td>
<td>1400903</td>
<td>other</td>
<td>61.914707870566346</td>
</tr>
</tbody>
</table>
<p>We can see that the assembly <code>REL606_evolved_2</code> had two inversions, one across the terminus and one across the origin, while <code>REL606_evolved_1</code> had only one inversion across the terminus. The symmetry of each of these inversions is described by the <code>symmetry_percent</code> field, which is the percent of the total length of the inversion that is in the longer arm of the inversion. This value is 50 for a perfectly symmetric inversion and becomes larger as the inversion becomes more asymmetric. </p></div>
            </div>
        </div>

        <footer class="col-md-12">
            <hr>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script src="../js/bootstrap.bundle.min.js"></script>
        <script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="../js/base.js"></script>
        <script src="../search/main.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
            </div>
            <div class="modal-body">
                <p>From here you can search these documents. Enter your search terms below.</p>
                <form>
                    <div class="form-group">
                        <input type="search" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results" data-no-results-text="No results found"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
