<!DOCTYPE html>
<html lang="en" data-bs-theme="light">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        
        
        
        <link rel="shortcut icon" href="../img/favicon.ico">
        <title>Usage - seabreeze</title>
        <link href="../css/bootstrap.min.css" rel="stylesheet">
        <link href="../css/fontawesome.min.css" rel="stylesheet">
        <link href="../css/brands.min.css" rel="stylesheet">
        <link href="../css/solid.min.css" rel="stylesheet">
        <link href="../css/v4-font-face.min.css" rel="stylesheet">
        <link href="../css/base.css" rel="stylesheet">
        <link id="hljs-light" rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" >
        <link id="hljs-dark" rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github-dark.min.css" disabled>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
        <script>hljs.highlightAll();</script> 
    </head>

    <body>
        <div class="navbar fixed-top navbar-expand-lg navbar-dark bg-primary">
            <div class="container">
                <a class="navbar-brand" href="..">seabreeze</a>
                <!-- Expander button -->
                <button type="button" class="navbar-toggler" data-bs-toggle="collapse" data-bs-target="#navbar-collapse" aria-controls="navbar-collapse" aria-expanded="false" aria-label="Toggle navigation">
                    <span class="navbar-toggler-icon"></span>
                </button>

                <!-- Expanded navigation -->
                <div id="navbar-collapse" class="navbar-collapse collapse">
                        <!-- Main navigation -->
                        <ul class="nav navbar-nav">
                            <li class="nav-item">
                                <a href=".." class="nav-link">Home</a>
                            </li>
                            <li class="nav-item">
                                <a href="../installation/" class="nav-link">Installation</a>
                            </li>
                            <li class="nav-item">
                                <a href="./" class="nav-link active" aria-current="page">Usage</a>
                            </li>
                            <li class="nav-item">
                                <a href="../output/" class="nav-link">Output</a>
                            </li>
                            <li class="nav-item">
                                <a href="../tutorial/" class="nav-link">Tutorial</a>
                            </li>
                            <li class="nav-item">
                                <a href="../workflows/" class="nav-link">Workflows</a>
                            </li>
                        </ul>

                    <ul class="nav navbar-nav ms-md-auto">
                        <li class="nav-item">
                            <a href="#" class="nav-link" data-bs-toggle="modal" data-bs-target="#mkdocs_search_modal">
                                <i class="fa fa-search"></i> Search
                            </a>
                        </li>
                            <li class="nav-item">
                                <a rel="prev" href="../installation/" class="nav-link">
                                    <i class="fa fa-arrow-left"></i> Previous
                                </a>
                            </li>
                            <li class="nav-item">
                                <a rel="next" href="../output/" class="nav-link">
                                    Next <i class="fa fa-arrow-right"></i>
                                </a>
                            </li>
                    </ul>
                </div>
            </div>
        </div>

        <div class="container">
            <div class="row">
                    <div class="col-md-3"><div class="navbar-expand-md bs-sidebar hidden-print affix" role="complementary">
    <div class="navbar-header">
        <button type="button" class="navbar-toggler collapsed" data-bs-toggle="collapse" data-bs-target="#toc-collapse" title="Table of Contents">
            <span class="fa fa-angle-down"></span>
        </button>
    </div>

    
    <div id="toc-collapse" class="navbar-collapse collapse card bg-body-tertiary">
        <ul class="nav flex-column">
            
            <li class="nav-item" data-bs-level="1"><a href="#preparing-input" class="nav-link">Preparing input</a>
              <ul class="nav flex-column">
            <li class="nav-item" data-bs-level="2"><a href="#run-analyzing-one-ancestor-assembly-pair" class="nav-link">run: Analyzing one ancestor-assembly pair</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#batch-analyzing-multiple-samples-at-once" class="nav-link">batch: Analyzing multiple samples at once</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#valid-workflows" class="nav-link">Valid workflows</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#options" class="nav-link">Options</a>
              <ul class="nav flex-column">
              </ul>
            </li>
            <li class="nav-item" data-bs-level="2"><a href="#additional-snakemake-argument" class="nav-link">Additional Snakemake argument</a>
              <ul class="nav flex-column">
              </ul>
            </li>
              </ul>
            </li>
        </ul>
    </div>
</div></div>
                    <div class="col-md-9" role="main">

<h1 id="preparing-input">Preparing input</h1>
<p>Assuming you have <a href="../installation/">completed installation</a>, you can now prepare input files for analysis. At its core, <em>seabreeze</em> is designed for bacterial resequencing experiments, and therefore predicts mutations between reference-query pairs, where the reference sequence is referred to as the <em>ancestor</em> and the query sequence being compared to is called the <em>assembly</em>. This nomenclature is use throughout the documentation and the software.
<em>seabreeze</em> requires FASTA files for all of the bacterial genomes being analysed. These FASTA files should only contain a single contig for the full-length genome. It should not contain additional entries for plasmids. <em>seabreeze</em> works best with genome assemblies generated from long-read sequencing as these can capture structural variants that occur between repeat sequences. It also does not work well for comparing genomes with high sequence/ phylogenetic divergence (for example between different bacterial species).</p>
<p><em>seabreeze</em> has two modes: 
- <code>run</code>:  Analyzing one ancestor-assembly pair
- <code>batch</code>: Analyzing multiple samples at once</p>
<p>Both modes run the same workflows, but require input to be supplied differently. </p>
<h2 id="run-analyzing-one-ancestor-assembly-pair"><code>run</code>: Analyzing one ancestor-assembly pair</h2>
<pre><code>Usage: seabreeze run [OPTIONS] [WORKFLOW] [SNAKEMAKE_ARGS]...

Options:
  -d, --dir PATH     Set working directory. Default is current working
                     directory
  --ancestor PATH    Path to FASTA file of the ancestor or reference sequence
  --assembly PATH    Path to FASTA file of the assembly or query sequence
  --threads INTEGER  Resources to be used.  [default: 4]
  --ori TEXT         Sequence of the origin. Must use IUPAC bases, and be
                     unique in both the ancestor and assembly.  Required for predict_replichore_balance or run_all workflows
  --dif TEXT         Sequence of the terminus. Must use standard IUPAC bases,
                     and be unique in both the ancestor and assembly. Required for predict_replichore_balance or run_all workflows
  --masked           Mask insertion sequences
  -h, --help         Show this message and exit.

</code></pre>
<ul>
<li>All input is directly supplied via the command line. No set up required.</li>
</ul>
<h2 id="batch-analyzing-multiple-samples-at-once"><code>batch</code>: Analyzing multiple samples at once</h2>
<pre><code>Usage: seabreeze batch [OPTIONS] [WORKFLOW] [SNAKEMAKE_ARGS]...

  Determine snakemake command to run based on supplied args

Options:
  -d, --dir PATH     Set working directory. Default is current working
                     directory
  -b, --data PATH    Path to data.csv file. Default is data.csv in current
                     working dir
  --oridif PATH      Path to csv file with sequences for the origin and
                     terminus. Default is ori_dif_coords.csv in the current dir.
                     Required for predict_replichore_balance or run_all workflows
  --threads INTEGER  Resources to be used.  [default: 4]
  --masked           Mask insertion sequences
  -h, --help         Show this message and exit.

</code></pre>
<p><code>batch</code> mode requires the following set up:</p>
<ul>
<li><code>data.csv</code>: The pairwise comparisons to be performed are specified in a csv file called <code>data.csv</code> , whose path is supplied via the <code>--data</code> option. This file has only two columns <code>ancestor</code> and <code>assembly</code>, in a standard csv format. Populate the csv file with the ancestor-assembly pairs you wish to analyse. For example:</li>
</ul>
<table>
<thead>
<tr>
<th>assembly</th>
<th>ancestor</th>
</tr>
</thead>
<tbody>
<tr>
<td>genome1</td>
<td>genome2</td>
</tr>
<tr>
<td>genome3</td>
<td>genome4</td>
</tr>
</tbody>
</table>
<ul>
<li><code>02_genomes</code>: A subdirectory of the specified working directory in the <code>--dir</code> option. Place the corresponding FASTA files for these genomes in <code>02_genomes</code>. <em>seabreeze</em> expects that these file names to match, and contain the extension <code>.fasta</code>. For example the corresponding working directory should look like this</li>
</ul>
<pre><code>&lt;my_working_dir&gt;
|
|---02_genomes/
|   |
|   |---genome1.fasta
|   |---genome2.fasta
|   |---genome3.fasta
|   |---genome4.fasta
</code></pre>
<p>Note that many assemblies can use the same ancestor, but the same assembly cannot use many ancestors:
For example, this is fine:</p>
<table>
<thead>
<tr>
<th>assembly</th>
<th>ancestor</th>
</tr>
</thead>
<tbody>
<tr>
<td>genome1</td>
<td>genome2</td>
</tr>
<tr>
<td>genome3</td>
<td>genome2</td>
</tr>
</tbody>
</table>
<p>This will cause a fatal error:</p>
<table>
<thead>
<tr>
<th>assembly</th>
<th>ancestor</th>
</tr>
</thead>
<tbody>
<tr>
<td>genome1</td>
<td>genome2</td>
</tr>
<tr>
<td>genome1</td>
<td>genome3</td>
</tr>
</tbody>
</table>
<p>It is also a good idea to set up a comparison of the ancestor genomes to themselves. This can be a quality control step, as we do not expect the pipeline to predict any mutations in a genome relative to itself. The following table demonstrates an implementation of this:</p>
<table>
<thead>
<tr>
<th>assembly</th>
<th>ancestor</th>
</tr>
</thead>
<tbody>
<tr>
<td>genome1</td>
<td>genome2</td>
</tr>
<tr>
<td>genome3</td>
<td>genome4</td>
</tr>
<tr>
<td>genome2</td>
<td>genome2</td>
</tr>
<tr>
<td>genome4</td>
<td>genome4</td>
</tr>
</tbody>
</table>
<hr />
<h2 id="valid-workflows">Valid workflows</h2>
<ul>
<li><code>analyse_genome_sizes</code></li>
<li><code>predict_IS_elements</code></li>
<li><code>predict_structural_variants</code></li>
<li><code>predict_replichore_balance</code></li>
<li><code>predict_SV_mechanism</code></li>
<li><code>annotate_SV_regions</code></li>
<li><code>run_all</code></li>
</ul>
<p>For detailed information on each workflow, please see <a href="../workflows/">the workflows page</a></p>
<hr />
<h2 id="options">Options</h2>
<p>These options can be </p>
<ul>
<li><code>--masked</code>: Mask insertion sequences with NNNs. Masking repeats can help detect structural mutations.</li>
</ul>
<hr />
<h2 id="additional-snakemake-argument">Additional Snakemake argument</h2>
<p>Any valid Snakemake argument can be passed at the end of the command, but the following are the most commonly used options</p>
<ul>
<li><code>-np</code>: This flag performs a 'dry-run' of the command i.e. it checks to see that all of the input files necessary for that command are present, or can be generated by <em>seabreeze</em>. It can helpful to run this before a computationally expensive step such as predicting IS elements, or predicting SVs.</li>
</ul>
<p>If a command was terminated before it was completed, Snakemake may prompt you to use these flags when you run the same command again:</p>
<ul>
<li><code>--unlock</code>:  Snakemake will "lock" a directory if all of the output files were not completed before the processes were terminated. Use this flag to unlock the directory.</li>
<li>
<p><code>--rerun-incomplete</code>:  To allow Snakemake to remake any output files that were not completed in the prior run.</p>
</li>
<li>
<p><code>--conda-frontend conda</code>: Some users have reported a <code>CreatedCondaEnvironmentException</code> that arises when trying to run any of the Snakemake workflows (see below). Appending this flag should address this issue. </p>
</li>
</ul>
<pre><code>CreateCondaEnvironmentException:
The 'mamba' command is not available in the shell /usr/bin/bash that will be used by Snakemake. You have to ensure that it is in your PATH, e.g., first activating the conda base environment with `conda activate base`.The mamba package manager (https://github.com/mamba-org/mamba) is a fast and robust conda replacement. It is the recommended way of using Snakemake's conda integration. It can be installed with `conda install -n base -c conda-forge mamba`. If you still prefer to use conda, you can enforce that by setting `--conda-frontend conda`.
</code></pre>
<p>Another similar conda exception has also been reported, and appending this flag resolves the issue.</p>
<pre><code>error    libmamba Non-conda folder exists at prefix
critical libmamba Aborting
</code></pre></div>
            </div>
        </div>

        <footer class="col-md-12">
            <hr>
            <p>Documentation built with <a href="https://www.mkdocs.org/">MkDocs</a>.</p>
        </footer>
        <script src="../js/bootstrap.bundle.min.js"></script>
        <script>
            var base_url = "..",
                shortcuts = {"help": 191, "next": 78, "previous": 80, "search": 83};
        </script>
        <script src="../js/base.js"></script>
        <script src="../search/main.js"></script>

        <div class="modal" id="mkdocs_search_modal" tabindex="-1" role="dialog" aria-labelledby="searchModalLabel" aria-hidden="true">
    <div class="modal-dialog modal-lg">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="searchModalLabel">Search</h4>
                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
            </div>
            <div class="modal-body">
                <p>From here you can search these documents. Enter your search terms below.</p>
                <form>
                    <div class="form-group">
                        <input type="search" class="form-control" placeholder="Search..." id="mkdocs-search-query" title="Type search term here">
                    </div>
                </form>
                <div id="mkdocs-search-results" data-no-results-text="No results found"></div>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div><div class="modal" id="mkdocs_keyboard_modal" tabindex="-1" role="dialog" aria-labelledby="keyboardModalLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <h4 class="modal-title" id="keyboardModalLabel">Keyboard Shortcuts</h4>
                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
            </div>
            <div class="modal-body">
              <table class="table">
                <thead>
                  <tr>
                    <th style="width: 20%;">Keys</th>
                    <th>Action</th>
                  </tr>
                </thead>
                <tbody>
                  <tr>
                    <td class="help shortcut"><kbd>?</kbd></td>
                    <td>Open this help</td>
                  </tr>
                  <tr>
                    <td class="next shortcut"><kbd>n</kbd></td>
                    <td>Next page</td>
                  </tr>
                  <tr>
                    <td class="prev shortcut"><kbd>p</kbd></td>
                    <td>Previous page</td>
                  </tr>
                  <tr>
                    <td class="search shortcut"><kbd>s</kbd></td>
                    <td>Search</td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div class="modal-footer">
            </div>
        </div>
    </div>
</div>

    </body>
</html>
